v1.11 backports 2022-10-03

christarazi opened the pull request:
* #20721 -- Improve dnsproxy ipcache handling, metrics (@joestringer)
* #20833 -- ipcache: Fix lock leak (@joestringer)
* #21565 -- ipcache: Fix metadata access from CIDR allocation (@joestringer)
  * Minor conflicts. I also found one additional corner case locking issue which is fixed here. I'll send a separate fix to master.
  * https://github.com/cilium/cilium/pull/21653 -- ipcache: Release metadata mutex in loop error condition (@joestringer)
* #21629 -- Fix agent deadlock caused by frequent kube-apiserver IP recycling (@joestringer)
  * Conflicts: Use of *net.IPNet and string instead of netip.Prefix on v1.12 branch. Had to adapt some code to a global IPIdentityCache and using functions instead of methods for release.

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 20721 20833 21565 21629 21653; do contrib/backporting/set-labels.py $pr done 1.11; done
```

joestringer committed:
ipcache: Fix race in identity/ipcache release

[ upstream commit 40e13ea2a5a944a45761fc433c4c971536957f4b ]

[ Backporter's notes: Conflicts were mostly to convert the methods on
               (ipc *IPIdentityCache) to the global IPIdentityCache.
               Also included a lock leak fix (9238841, "ipcache: Fix
               lock leak") not in the original commit. ]

Create a critical section for identity release + removal from ipcache.
Otherwise, it's possible to trigger the following race condition:

    Goroutine 1                | Goroutine 2
    ---------------------------+--------------------------------------
    releaseCIDRIdentities()    | AllocateCIDRs()
    -> Release(..., id, ...)   |
                               | -> allocate(...)
                               | -> ipc.UpsertGeneratedIdentities(...)
    -> ipc.deleteLocked(...)   |

In this case, the expectation from Goroutine 2 is that a new identity
is allocated and that identity is inserted into the ipcache, but the
result is that the identity is allocated but the ipcache entry is
missing. This is partly because the identity released in goroutine 1 is
different from the newly allocated identity in goroutine 2, however
goroutine 1 will delete the ipcache entry based on the prefix and not
the identity. Therefore it's possible for goroutine 1 to delete the
ipcache entry corresponding to the identity allocated in goroutine 2.

Note that for balancing the upsert / release, we perhaps should cover
the entire allocation + ipcache push in Upsert() with the same locking.
However, on upsert there is an optional feature of the API to defer the
ipcache upsert to a later point, governed by the caller. There is
currently no way to extend the locking over that much longer time
period, so we only cover the allocation step there. This should still be
safe, as one of the following cases should occur:

    Goroutine 1                | Goroutine 2
    ---------------------------+--------------------------------------
    Lock                       |
    Release()                  |
    deleteLocked()             |
    Unlock                     |
                               | Lock
                               | ipc.allocate()
                               | Unlock
                               | .... (repeat below)
                               | Lock
                               | Upsert
                               | Unlock

    Goroutine 1                | Goroutine 2
    ---------------------------+--------------------------------------
                               | Lock
                               | ipc.allocate() (increment refcount)
                               | Unlock
    Lock                       |
    Release()                  |
    (no deleteLocked()() due   |
     to refcount from (2))     |
    Unlock                     |
                               | .... (repeat below)
                               | Lock
                               | Upsert
                               | Unlock

Found by code inspection.

Suggested-by: Jarno Rajahalme <jarno@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
fqdn: Upsert all identities to ipcache

[ upstream commit e6ad7438357da93e5c5dbf823e71ae349adde61d ]

[ Backporter's notes: Conflicts were mostly to convert the methods on
               (ipc *IPIdentityCache) to the global IPIdentityCache. ]

Previously, the logic would only upsert identities into the IPCache if
the identity was newly allocated. Logically this makes sense, as the
relationship between a CIDR identity and the ipcache should be tightly
coupled. However, we have observed in some user environments that
ipcache entries may end up being removed from the datapath and the
corresponding identity would remain allocated in userspace. As a result,
the next time a DNS request arrives which intends to make use of that
identity for subsequent connection attempts, it would not populate the
ipcache with the identity, leading to packet loss on the connection
allowed by ToFQDNs policy.

In order to mitigate this issue, ensure that all identities used in DNS
responses are populated into the datapath, and track a metric for any
cases where this occurs for identities that we expect to already be
present in the IPCache. This way, active issues should be mitigated,
but we also still have a way to detect whether this mitigation is
necessary and whether we need to further investigate the root cause of
this issue.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache: Add metrics for upsert/delete/recover

[ upstream commit 044cd8f23cc531d7441127a15832c39fc43d159d ]

These errors, total metrics will help users and developers to gather
understanding about ipcache operations at runtime. One specific
"recover" error that will occur at runtime will be measured in an
upcoming commit. This is the primary motivation for introducing these
metrics.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache: Fix metadata access from CIDR allocation

[ upstream commit 021ea42e6f8cfbb139286db88c4d5e2f84dac93f ]

[ Backporter's notes: Fixed release of metadata lock in error case.
		      Additionally, adapt to static IPIdentityCache. ]

The locking order for metadata -> IPCache should be first grabbing the
metadata lock then the IPCache lock, according to the documentation in
the metadata structure. Correct this lock ordering to conform to the
documented pattern.

This sort of improper lock ordering could theoretically cause a deadlock
if for instance the label injector ran at the same time as this function
is called.

Found by inspection.

Fixes: 40e13ea2a5a9 ("ipcache: Fix race in identity/ipcache release")
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on pkg/ipcache/cidr.go:
```diff
@@ -56,6 +57,7 @@ func AllocateCIDRs(
 		newlyAllocatedIdentities = map[string]*identity.Identity{}
 	}
 
+	IPIdentityCache.Lock()
```
Yeah this part could trigger the bug fixed in https://github.com/cilium/cilium/pull/21565.

joestringer committed:
ipcache: Defer CIDR identity release and deletion

[ upstream commit 4afbcc38ce91118f1cb4a9f715a268775cdf7f6e ]

[ Backporter's notes: Switch from netip to string internally, since the
                      older branch doesn't use netip in ipcache.
		      Additionally, adapt to static IPIdentityCache. ]

A user reports the following deadlock in with lock chain[^1] and
lock chain[^2].

Lock chain 2 is at a high level trying to implement the new network
policy for an endpoint. As part of this step, it calculates a new
policy, then attempts to switch the policy pointer over to the new
policy, and then garbage collect references & objects in the old policy.
When garbage collecting the old policy objects, it is going into the
selectorcache and identifying that certain identities are no longer used
by the selector and cleaning up those identities. Hence we end up in the
ReleaseCIDRIdentitiesByID() call and attempting to grab the ipcache
lock.

Commit 40e13ea attempted to mitigate a bug where interleaved ordering of
identity reference counting and ipcache updates could lead to
inconsistent internal state (=> packet drops), but the core idea there
was that these two operations (identity refcount update + ipcache
update) must happen together in the same critical section, but the exact
timing when these two operations occur is not particularly important. I
think they should happen in a canonical order, but when that order is
iterated is not critical. Given that overall agent state is eventually
consistent, cleanup of these ipcache entries can happen at any
subsequent time. The sooner the better for sure, but I don't think that
there are any hard constraints on this.

If lock chain 2 was not holding the endpoint lock while releasing these
resources, then this deadlock would not occur. So, this commit proposes
to add a new queue + GC goroutine in the ipcache that handles the
release of these CIDR identities out-of-band so that they can be cleaned
up while not holding the endpoint lock. Then, from the endpoint policy
generation perspective it can continue with the remaining endpoint
policy calculation / datapath updates, return, and then that will allow
lock chain 1 to proceed since it's waiting for all endpoints to
regenerate their policies, then once that completes & unlocks the lock,
this new dedicated goroutine can do this cleanup.

[1]: Lock chain 1: ipc.Lock() -> e.lockAlive()
     When removeLabelsFromIPs is triggered through an EP change, it performs
     ipc.Lock() first, then collects information about added/deleted
     identities in a loop and then calls ipc.UpdatePolicyMaps(...), which
     then calls UpdatePolicyMaps(...) from the EndpointManager. The call
     stack of this is:

```
goroutine 237 [semacquire, 417 minutes]:
sync.runtime_Semacquire(0xc00038d880?)
	/usr/local/go/src/runtime/sema.go:56 +0x25
sync.(*WaitGroup).Wait(0xc000398700?)
	/usr/local/go/src/sync/waitgroup.go:136 +0x52
github.com/cilium/cilium/pkg/ipcache.(*IPCache).UpdatePolicyMaps(0xc00083c980, {0x3461140, 0xc000128008}, 0xc?, 0xc0024410e0)
	/go/src/github.com/cilium/cilium/pkg/ipcache/metadata.go:235 +0xc7
github.com/cilium/cilium/pkg/ipcache.(*IPCache).removeLabelsFromIPs(0xc00083c980, 0xc0013d5778?, {0x2f2a9f6, 0xf})
	/go/src/github.com/cilium/cilium/pkg/ipcache/metadata.go:414 +0x7c5
github.com/cilium/cilium/pkg/ipcache.(*IPCache).RemoveLabelsExcluded(0xc00083c980, 0xc0000d5c50, 0xc000bf41d8?, {0x2f2a9f6, 0xf})
	/go/src/github.com/cilium/cilium/pkg/ipcache/metadata.go:328 +0x1ab
github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).handleKubeAPIServerServiceEPChanges(0xc000a91d40, 0xc001a739b0?)
	/go/src/github.com/cilium/cilium/pkg/k8s/watchers/endpoint.go:135 +0x5b
github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).addKubeAPIServerServiceEPSliceV1(0x1861426?, 0xc0020f2c30)
	/go/src/github.com/cilium/cilium/pkg/k8s/watchers/endpoint_slice.go:205 +0x452
github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).updateK8sEndpointSliceV1(0xc000a91d40, 0xc0020f2c30?, 0xc0020f2c30?)
	/go/src/github.com/cilium/cilium/pkg/k8s/watchers/endpoint_slice.go:178 +0x69
github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).endpointSlicesInit.func2({0x2ebcec0?, 0xc0020f3a00?}, {0x2ebcec0, 0xc0020f2c30})
	/go/src/github.com/cilium/cilium/pkg/k8s/watchers/endpoint_slice.go:71 +0x125
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:239
github.com/cilium/cilium/pkg/k8s/informer.NewInformerWithStore.func1({0x2a48d00?, 0xc000a9b5f0?})
	/go/src/github.com/cilium/cilium/pkg/k8s/informer/informer.go:103 +0x2fe
k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0xc000c5a960, 0xc00013b240)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:554 +0x566
k8s.io/client-go/tools/cache.(*controller).processLoop(0xc000eff710)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:184 +0x36
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x40d645?)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x102fe25?, {0x3436f20, 0xc000b54b10}, 0x1, 0xc0005e06c0)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000eff778?, 0x3b9aca00, 0x0, 0x40?, 0x7f72ce949c00?)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90
k8s.io/client-go/tools/cache.(*controller).Run(0xc000eff710, 0xc0005e06c0)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:155 +0x2c5
created by github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).endpointSlicesInit
	/go/src/github.com/cilium/cilium/pkg/k8s/watchers/endpoint_slice.go:156 +0x759
UpdatePolicyMaps will then start a goroutine for each EP and just return a WaitGroup, so that the caller can wait on it. The goroutines then end up in calling locking the EP lock via lockAlive().
```

[2]: Lock chain 2: e.lockAlive() -> ipc.Lock()
     If at the same time a EndpointRegenerationEvent is handled, it
     might end up locking the Endpoint locks first and then (through
     a quite deep stack) end up calling ipc.Lock() in
     IPCache.releaseCIDRIdentities(), which will then cause a
     deadlock. The stack trace looks like this:

```
goroutine 455 [select, 417 minutes]:
golang.org/x/sync/semaphore.(*Weighted).Acquire(0xc000842050, {0x3461140, 0xc000128000}, 0x40000000)
	/go/src/github.com/cilium/cilium/vendor/golang.org/x/sync/semaphore/semaphore.go:60 +0x345
github.com/cilium/cilium/pkg/lock.(*SemaphoredMutex).Lock(...)
	/go/src/github.com/cilium/cilium/pkg/lock/semaphored_mutex.go:30
github.com/cilium/cilium/pkg/ipcache.(*IPCache).Lock(...)
	/go/src/github.com/cilium/cilium/pkg/ipcache/ipcache.go:121
github.com/cilium/cilium/pkg/ipcache.(*IPCache).releaseCIDRIdentities(0xc00083c980, {0x3461178, 0xc000445680}, 0x0?)
	/go/src/github.com/cilium/cilium/pkg/ipcache/cidr.go:203 +0x85
github.com/cilium/cilium/pkg/ipcache.(*IPCache).ReleaseCIDRIdentitiesByID(0xc00083c980, {0x3461178, 0xc000445680}, {0x0, 0x0, 0x1bf08eb000?})
	/go/src/github.com/cilium/cilium/pkg/ipcache/cidr.go:265 +0x497
github.com/cilium/cilium/daemon/cmd.cachingIdentityAllocator.ReleaseCIDRIdentitiesByID(...)
	/go/src/github.com/cilium/cilium/daemon/cmd/identity.go:118
github.com/cilium/cilium/pkg/policy.(*SelectorCache).releaseIdentityMappings(0xc000398700, {0x0, 0x0, 0x0})
	/go/src/github.com/cilium/cilium/pkg/policy/selectorcache.go:556 +0x9d
github.com/cilium/cilium/pkg/policy.(*SelectorCache).RemoveSelectors(0xc000398700, {0xc001a00210, 0x1, 0x0?}, {0x3429f40, 0xc003d42600})
	/go/src/github.com/cilium/cilium/pkg/policy/selectorcache.go:961 +0xc5
github.com/cilium/cilium/pkg/policy.(*L4Filter).removeSelectors(0xc003d42600, 0xc001c02518?)
	/go/src/github.com/cilium/cilium/pkg/policy/l4.go:627 +0x185
github.com/cilium/cilium/pkg/policy.(*L4Filter).detach(0xb?, 0x0?)
	/go/src/github.com/cilium/cilium/pkg/policy/l4.go:634 +0x1e
github.com/cilium/cilium/pkg/policy.L4PolicyMap.Detach(...)
	/go/src/github.com/cilium/cilium/pkg/policy/l4.go:804
github.com/cilium/cilium/pkg/policy.(*L4Policy).Detach(0xc0048b0180, 0x1?)
	/go/src/github.com/cilium/cilium/pkg/policy/l4.go:1013 +0x7b
github.com/cilium/cilium/pkg/policy.(*selectorPolicy).Detach(...)
	/go/src/github.com/cilium/cilium/pkg/policy/resolve.go:104
github.com/cilium/cilium/pkg/policy.(*cachedSelectorPolicy).setPolicy(0xc000398770?, 0xc000c275c0?)
	/go/src/github.com/cilium/cilium/pkg/policy/distillery.go:188 +0x3b
github.com/cilium/cilium/pkg/policy.(*PolicyCache).updateSelectorPolicy(0xc00098a090, 0xc000c275c0)
	/go/src/github.com/cilium/cilium/pkg/policy/distillery.go:124 +0x195
github.com/cilium/cilium/pkg/policy.(*PolicyCache).UpdatePolicy(...)
	/go/src/github.com/cilium/cilium/pkg/policy/distillery.go:153
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regeneratePolicy(0xc000e47c00)
	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:230 +0x22b
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).runPreCompilationSteps(0xc000e47c00, 0xc0015cf400)
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:809 +0x2c5
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF(0xc000e47c00, 0xc0015cf400)
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:579 +0x189
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate(0xc000e47c00, 0xc0015cf400)
	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:398 +0x7a5
github.com/cilium/cilium/pkg/endpoint.(*EndpointRegenerationEvent).Handle(0xc005584d10, 0xc0003dae80?)
	/go/src/github.com/cilium/cilium/pkg/endpoint/events.go:53 +0x325
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run.func1()
	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:245 +0x13b
sync.(*Once).doSlow(0x1?, 0x442205?)
	/usr/local/go/src/sync/once.go:68 +0xc2
sync.(*Once).Do(...)
	/usr/local/go/src/sync/once.go:59
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run(0x0?)
	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:233 +0x45
created by github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).Run
	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:229 +0x76
```

Thanks to John Watson, Alexander Block, and Chris Tarazi for their
reports and assistance in digging to the bottom of this issue.

CC: Joe Stringer <joe@cilium.io>
Fixes: 40e13ea2a5a9 ("ipcache: Fix race in identity/ipcache release")
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer reviewed (approved):
Feels a bit weird to be approving my own changes, but the backports lgtm... I did take another fresh look through and I understand Chris took a look back over the latest ones as well.

I think this is good to merge.
